core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -338,13 +340,19 @@ public DataWriteBuilder withSortOrder(SortOrder newSortOrder) {
       return this;
     }
 
+    public DataWriteBuilder withSchemaId(int newSchemaId) {
+      this.schemaId = newSchemaId;
```

Comment:
Why isn't this taken from the schema that is passed in?

Reply:
I see the schema passed into the `Avro`/`Parquet`/`ORC` is the one converted from Spark/Flink schema. And that schema has the default ID `-1`.

Reply:
Where is this happening?

Reply:
You could find it in the following code. The `new Schema(struct.fields())` will create the schema with the default schema ID 0.

```java
  Schema writeSchema = validateOrMergeWriteSchema(table, dsSchema, writeConf);

  // validateOrMergeWriteSchema
  private static Schema validateOrMergeWriteSchema(Table table, StructType dsSchema, SparkWriteConf writeConf) {
    Schema writeSchema;
    if (writeConf.mergeSchema()) {
      ...
    } else {
      writeSchema = SparkSchemaUtil.convert(table.schema(), dsSchema);
      TypeUtil.validateWriteSchema(
          table.schema(), writeSchema, writeConf.checkNullability(), writeConf.checkOrdering());
    }

    return writeSchema;
  }

  // SparkSchemaUtil.convert
  public static Schema convert(Schema baseSchema, StructType sparkType) {
    // convert to a type with fresh ids
    Types.StructType struct = SparkTypeVisitor.visit(sparkType, new SparkTypeToType(sparkType)).asStructType();
    // reassign ids to match the base schema
    Schema schema = TypeUtil.reassignIds(new Schema(struct.fields()), baseSchema);
    // fix types that can't be represented in Spark (UUID and Fixed)
    return SparkFixupTypes.fixup(schema, baseSchema);
  }
```